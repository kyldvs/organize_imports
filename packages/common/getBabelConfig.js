/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

const DEFAULT_CONFIG = {
  sourceType: 'module',
  plugins: [
    'syntax-async-functions',
    'syntax-flow',
    'syntax-jsx',
    'syntax-object-rest-spread',
    'syntax-trailing-function-commas',
  ],
};

export default function getBabelConfig(config) {
  if (!config) {
    return DEFAULT_CONFIG;
  }
  const result = {};
  const allKeys = [].concat(Object.keys(config), Object.keys(DEFAULT_CONFIG));
  allKeys.forEach(key => {
    if (config[key]) {
      if (Array.isArray(DEFAULT_CONFIG[key])) {
        result[key] = [].concat(DEFAULT_CONFIG[key], config[key]);
      } else {
        result[key] = config[key];
      }
    } else {
      result[key] = DEFAULT_CONFIG[key];
    }
  });
  return result;
};
