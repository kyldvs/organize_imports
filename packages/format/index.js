/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

import getRequireKind from '../common/getRequireKind';
import match from '../common/match';

const transforms = [
  sortDestructures,
  sortSpecifiers,
];

// Only valid requires should be passed into this function
const config = [
  // Type Imports
  {
    type: 'ImportDeclaration',
    importKind: 'type',
  },

  // Side Effects
  {
    type: 'ExpressionStatement',
  },

  // TitleCase
  node => {
    if (node.type === 'ImportDeclaration') {
      return isCapitalized(node.specifiers[0].local.name);
    }
    if (node.type === 'VariableDeclaration') {
      const id = node.declarations[0].id;
      if (id.type === 'Identifier') {
        return isCapitalized(id.name);
      }
      if (id.type === 'ObjectPatern') {
        if (id.key.type === 'ObjectProperty') {
          return isCapitalized(id.key.name);
        }
        if (id.key.type === 'RestProperty') {
          // We treat rest arguments as "camelCase"
          return false;
        }
      }
    }
    return false;
  },

  // camelCase
  node => {
    if (node.type === 'ImportDeclaration') {
      return !isCapitalized(node.specifiers[0].local.name);
    }
    if (node.type === 'VariableDeclaration') {
      const id = node.declarations[0].id;
      if (id.type === 'Identifier') {
        return !isCapitalized(id.name);
      }
      if (id.type === 'ObjectPatern') {
        if (id.key.type === 'ObjectProperty') {
          return !isCapitalized(id.key.name);
        }
        if (id.key.type === 'RestProperty') {
          // We treat rest arguments as "camelCase"
          return true;
        }
      }
    }
    return false;
  },
];

export default function format(options) {
  return function babel_plugin({types: t}) {
    return {
      visitor: {
        Program(path) {

          const pre = [];
          const requires = [];
          const post = [];

          path.node.body.forEach(node => {
            if (getRequireKind(node)) {
              requires.push(node);
            } else {
              if (requires.length > 0) {
                post.push(node);
              } else {
                pre.push(node);
              }
            }
          });

          const transformed = requires.map(node => transforms.reduce(
            (node, fn) => fn(node),
            node,
          ));

          const groups = [];
          for (let i = 0; i < config.length; i++) {
            groups.push([]);
            transformed.forEach(node => {
              if (match(node, config[i])) {
                groups[i].push(node);
              }
            });
          }

          const resultRequires = [].concat(...groups);
          const resultBody = [].concat(pre, resultRequires, post);

          path.node.body = resultBody;
        },
      },
    };
  };
}

/**
 * This should only be passed import nodes, it will sort object patterns on the
 * left side of a require.
 */
function sortDestructures(node) {
  if (!match(node, {
    type: 'VariableDeclaration',
    declarations: [{
      type: 'VariableDeclarator',
      id: {
        type: 'ObjectPattern',
      },
    }],
  })) {
    return node;
  }

  node.declarations[0].id.properties.sort((p1, p2) => {
    // Rest properties always go at the end.
    if (p1.type === 'RestProperty') {
      return 1;
    }
    if (p2.type === 'RestProperty') {
      return -1;
    }

    // Sort simple properties by their identifier.
    if ([p1, p2].every(p => match(p, {
      type: 'ObjectProperty',
      key: {
        type: 'Identifier',
      },
    }))) {
      if (p1.key.name < p2.key.name) {
        return -1;
      } else if (p1.key.name > p2.key.name) {
        return 1;
      } else {
        return 0;
      }
    }

    // Not sure how to compare them otherwise. This should probable never
    // happen.
    return 0;
  });
}

/**
 * This should only be passed import nodes, it will sort specifiers in an
 * import.
 */
function sortSpecifiers(node) {
  return node;
}

function isCapitalized(str) {
  return str.charAt(0).toUpperCase() === str.charAt(0);
}
