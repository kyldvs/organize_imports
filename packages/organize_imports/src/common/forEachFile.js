/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

import fs from 'fs';
import path from 'path';

export default function forEachFile(dir, fn) {
  getFiles(dir, dir).forEach(x => fn(x));
}

function getFiles(firstFilePath, originalFilePath) {
  let arr = [];
  const filePaths = fs.readdirSync(originalFilePath);
  filePaths.forEach(fileName => {
    const filePath = path.join(originalFilePath, fileName);
    const stats = fs.statSync(filePath);
    if (stats.isDirectory()) {
      arr = [].concat(arr, getFiles(firstFilePath, filePath));
    } else if (stats.isFile()) {
      arr.push({
        name: fileName,
        relativeName: path.relative(firstFilePath, filePath),
        contents: fs.readFileSync(filePath, 'utf8'),
      });
    }
  });
  return arr;
}
