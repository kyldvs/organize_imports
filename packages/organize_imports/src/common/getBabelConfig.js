/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

const DEFAULT_CONFIG = {
  sourceType: 'module',
  plugins: [
    require('babel-plugin-syntax-async-functions'),
    require('babel-plugin-syntax-flow'),
    require('babel-plugin-syntax-jsx'),
    require('babel-plugin-syntax-object-rest-spread'),
    require('babel-plugin-syntax-trailing-function-commas'),
  ],
};

export default function getBabelConfig(config) {
  if (!config) {
    return DEFAULT_CONFIG;
  }
  const result = {};
  const allKeys = [].concat(Object.keys(config), Object.keys(DEFAULT_CONFIG));
  allKeys.forEach(key => {
    if (config[key]) {
      if (Array.isArray(DEFAULT_CONFIG[key])) {
        result[key] = [].concat(DEFAULT_CONFIG[key], config[key]);
      } else {
        result[key] = config[key];
      }
    } else {
      result[key] = DEFAULT_CONFIG[key];
    }
  });
  return result;
};
