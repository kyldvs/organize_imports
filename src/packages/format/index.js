/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

import RequireGroups from '../common/RequireGroups';

import getRequireIDs from '../common/getRequireIDs';
import isRequire from '../common/isRequire';
import match from '../common/match';
import reprintRequireGroups from '../common/reprintRequireGroups';

export default function format(options) {
  return function babel_plugin({types: t}) {
    return {
      visitor: {
        Program(path) {

          const pre = [];
          const requires = [];
          const post = [];

          // Partition all the body nodes, into pre-require-block,
          // require-block, and post-require-block.
          path.node.body.forEach(node => {
            if (isRequire(node)) {
              requires.push(node);
            } else {
              if (requires.length > 0) {
                post.push(node);
              } else {
                pre.push(node);
              }
            }
          });

          // Transform require nodes as appropriate.
          requires.forEach(node => {
            // TODO: Sort specifiers and destructures.
          });

          // Sort all the requires, this will prevent us needing to re-order
          // the groups later.
          requires.sort((n1, n2) => {
            const ids1 = getRequireIDs(n1);
            const ids2 = getRequireIDs(n2);

            // This should never happen.
            if (ids1.length === 0 || ids2.length === 0) {
              return 0;
            }

            for (let i = 0; i < ids1.length && i < ids2.length; i++) {
              if (ids1[i] < ids2[i]) {
                return -1;
              } else if (ids1[i] > ids2[i]) {
                return 1;
              }
            }

            // Since they are all equal to this point we can compare based on
            // the number of ids.
            if (ids1.length < ids2.length) {
              return -1;
            } else if (ids1.length > ids2.length) {
              return 1;
            } else {
              return 0;
            }
          });

          // Create an empty array for each group.
          const groups = RequireGroups.map(() => []);
          const unknowns = [];

          // For each require add it to the correct group.
          requires.forEach(node => {
            for (let i = 0; i < RequireGroups.length; i++) {
              if (RequireGroups[i].some(pattern => match(node, pattern))) {
                groups[i].push(node);
                return;
              }
            }

            // If we make it to here that's bad, but let's not drop the node.
            unknowns.push(node);
          });

          const newRequires = reprintRequireGroups(groups);

          // Sort each group.
          path.node.body = [... pre, ...newRequires, ...post];
        },
      },
    };
  };
}
