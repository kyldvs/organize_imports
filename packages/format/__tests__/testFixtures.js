/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

import format from '../index.js';
import * as babel from 'babel-core';
import forEachFile from '../../common/forEachFile';
import getBabelConfig from '../../common/getBabelConfig';
import path from 'path';

const FIXTURES_DIR = path.join(__dirname, '..', '__fixtures__');

forEachFile(FIXTURES_DIR, ({relativeName, contents}) => {
  test(relativeName, () => {
    expect(run(contents)).toMatchSnapshot();
  });
});

function run(code, options) {
  return babel.transform(code, getBabelConfig({
    plugins: [
      format(options),
    ],
  })).code;
}
