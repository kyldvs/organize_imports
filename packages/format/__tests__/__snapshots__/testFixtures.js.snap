exports[`test flow/removes_types_used_in_value_only.js 1`] = `
"
import type foo from \'foo\';

import type foo from \'foo\';foo();"
`;

exports[`test flow/removes_values_used_as_types_only.js 1`] = `
"
import foo from \'foo\';

function bar(): foo {}"
`;

exports[`test simple/one-used-one-unused.js 1`] = `
"const foo = require(\'foo\');
const bar = require(\'bar\');

foo();"
`;

exports[`test simple/will-be-empty.js 1`] = `"const foo = require(\'foo\');"`;
